Counting bloom filter adjustments

jowong4 opened the pull request:
This pull request includes minor adjustments suggested by Shaun. The only thing that is missing is assert_good which is from another repo. Do you want me to vendor that from abyss? or just leave the assert statements as is?

sjackman commented:
> The only thing that is missing is assert_good which is from another repo. Do you want me to vendor that from abyss?

Yes, please do copy it in. It's a useful function. It gives a much better error message. You can do that in a separate PR.

sjackman commented on CountingBloomFilter.hpp:
```diff
@@ -126,12 +117,11 @@ class CountingBloomFilter
 	static const uint32_t BloomFilter_VERSION = 2;
 	unsigned m_countThreshold;
 	unsigned m_bitsPerCounter = 8;
+	static constexpr const char* MAGIC_HEADER_STRING = "BLOOMCOU\0";
```
```suggestion
	static constexpr const char* MAGIC_HEADER_STRING = "BLOOMCOU";
```
The "\0" is implicit, so you don't need it here. As it is you would actually get two terminating zeros, the explicit one and the implicit one. Alternatively you can do
```c++
static constexpr const char MAGIC_HEADER_STRING[] = "BLOOMCOU";
```
which is similar but slightly different. One big difference is that `sizeof(MAGIC_HEADER_STRING)` of the `char*` type returns the size of a pointer, 8 bytes, whereas `sizeof(MAGIC_HEADER_STRING)` of the `char[]` returns the size of the string including the terminating "\0" (I believe, good to double check). I have a slight preference for `[]`.

sjackman replied:
I recommend that we use the same `MAGIC_HEADER_STRING` for both counting Bloom filters and non-counting Bloom filters, and in non-counting Bloom filters set `m_bitsPerCounter = 1`. That way the file format and serialization code can be shared between the two data structures. It avoids unnecessary code duplication and file format proliferation. I suggest we use a magic string of `BTLBloom` (eight characters).
cc @jowong4 @schutzekatze @JustinChu

sjackman commented on CountingBloomFilter.hpp:
```diff
@@ -319,25 +308,25 @@ CountingBloomFilter<T>::readHeader(FILE* fp)
-	char magic[9];
+	char magic[magicLen];
 	memcpy(magic, header.magic, 8);
 	magic[8] = '\0';
```
Delete these three lines.

jowong4 replied:
The reason for this code is I think is this: If you were to print out `header.magic` you get
`BLOOMCOU@' which will fail `strcmp(header.magic, MAGIC_HEADER_STRING)` because `MAGIC_HEADER_STRING` evaluate to`
BLOOMCOU`

sjackman commented on CountingBloomFilter.hpp:
```diff
@@ -126,12 +117,11 @@ class CountingBloomFilter
 	unsigned m_countThreshold;
 	unsigned m_bitsPerCounter = 8;
+	static constexpr const char* MAGIC_HEADER_STRING = "BLOOMCOU\0";
+	unsigned magicLen = strlen(MAGIC_HEADER_STRING) + 1;
```
```suggestion
```
Delete this line.

sjackman commented on CountingBloomFilter.hpp:
```diff
@@ -319,25 +308,25 @@ CountingBloomFilter<T>::readHeader(FILE* fp)
 	memcpy(magic, header.magic, 8);
 	magic[8] = '\0';
-	if (strcmp(magic, "BLOOMCOU")) {
+	if (strcmp(magic, MAGIC_HEADER_STRING)) {
```
```suggestion
	if (strcmp(header.magic, MAGIC_HEADER_STRING) != 0) {
```

sjackman replied:
```suggestion
	if (memcpy(magic, MAGIC_HEADER_STRING, strlen(MAGIC_HEADER_STRING)) != 0) {
```
As discussed in person.

jowong4 commented:
When I removed `using namespace std;` the code could still compile, so I read through the code and added `std::` to classes I know that requires the namespace. I'm not sure if I caught everything though. I have ran and execute the code and it seems to be fine.

sjackman commented:
Check that there's no `using namespace std;` in other header files in the code base.

sjackman commented:
> Check that there's no using namespace std; in other header files in the code base.

sjackman commented:
> Check that there's no using namespace std; in other header files in the code base.

If there are you, could address that issue in a separate PR.

sjackman commented on CountingBloomFilter.hpp:
```diff
@@ -5,49 +5,37 @@
 #include <cmath>
 #include <cstring>
 #include <fstream>
+#include <iostream>
```
In a header library, include `ostream` rather than `iostream`.

jowong4 replied:
If I were to use `ostream`,  I get  `error: no member named 'cerr' in namespace 'std' [clang-diagnostic-error] std::cerr << "Bloom Filter type does not match" << std::endl;`.

sjackman replied:
Yes, you're right. To use `std::cerr`, it needs to include `<iostream>`.

sjackman commented on CountingBloomFilter.hpp:
```diff
@@ -88,21 +76,21 @@ class CountingBloomFilter
 	struct FileHeader
 	{
-		char magic[8];
+		char magic[sizeof(uint64_t)];
```
```suggestion
		char magic[8];
```
I preferred the explicit `8` in this case. Alternatively, use a `static const unsigned MAGIC_LENGTH = 8;`

jowong4 replied:
So the reason I changed it in the first place was because clang-tidy said 8 was a magic number and the solution I came up with was that we choose 8 because it is the sizeof(uint64_t). 

 I also get an error `error: 'MAGIC_LENGTH' was not declared in this scope char magic[MAGIC_LENGTH];` When I try to declare that as a static variable and use it to create the char array.

sjackman replied:
This quick test worked for me:
```c++
class Hello {
	static const unsigned MAGIC_LENGTH = 8;
	char m_magic[MAGIC_LENGTH];
};
```

jowong4 replied:
The reason for the error I posted above is because the public FileHeader struct is defined before the private member initialization. I believe that there is no reason for the FileHeader to be public so I've moved it to private and then incorporated your changes.

sjackman commented on CountingBloomFilter.hpp:
```diff
@@ -88,21 +76,21 @@ class CountingBloomFilter
 	};
-	void readHeader(FILE* file);
-	void readFilter(const string& path);
+	void readHeader(FILE* fp);
```
I preferred the word `file` over the acronym `fp`.

sjackman replied:
I personally often use the variables `fin` and `fout` (though it is admittedly an abbreviation).

sjackman commented on CountingBloomFilter.hpp:
```diff
@@ -88,21 +76,21 @@ class CountingBloomFilter
+	void readFilter(const std::string& path);
 	void writeHeader(std::ostream& out) const;
-	void writeFilter(string const& path) const;
+	void writeFilter(std::string const& path) const;
```
```suggestion
	void writeFilter(const std::string& path) const;
```

sjackman commented on CountingBloomFilter.hpp:
```diff
@@ -5,49 +5,37 @@
-	  , m_nEntry(0)
-	  , m_tEntry(0)
-	  , m_countThreshold(0)
+	  : m_filter(nullptr)
```
Please initialize `m_filter` below as with the integer members and delete this line.

sjackman commented on CountingBloomFilter.hpp:
```diff
@@ -113,25 +101,21 @@ class CountingBloomFilter
 	// m_bitsPerCounter     : Number of bits per counter.
+	// MAGIC_HEADER_STRING  : Magic string used to identify the type of bloom filter.
 
 	T* m_filter;
```
```suggestion
	T* m_filter = nullptr;
```

sjackman commented on CountingBloomFilter.hpp:
```diff
@@ -113,25 +101,21 @@ class CountingBloomFilter
-	unsigned m_countThreshold;
+	unsigned m_countThreshold = 0;
 	unsigned m_bitsPerCounter = 8;
+	static constexpr const char* MAGIC_HEADER_STRING = "BLOOMCOU";
```
```suggestion
	static constexpr const char* MAGIC_HEADER_STRING = "BTLBloom";
```

sjackman commented on CountingBloomFilter.hpp:
```diff
@@ -254,61 +240,62 @@ CountingBloomFilter<T>::filtered_popcount() const
-  , m_nEntry(0)
-  , m_tEntry(0)
+CountingBloomFilter<T>::CountingBloomFilter(const std::string& path, unsigned countThreshold)
+  : m_filter(nullptr)
```
Delete this line and initialize at the declaration.

sjackman commented:
Thank you!
